### PR TITLE
Identify newline separated segments

### PR DESCRIPTION
**Pull request by sungshik**
This PR adds improved support for segment-based "gobbling up" tokens in begin/end patterns. To motivate, suppose we have the following grammar:

```
lexical String = @category="string" "\"" Char* "\"";

lexical Char
    = [a-z]
    | "\\" [\"]
    ;
```

For instance, `"asdf"` and `"foo\"bar"` should be fully highlighted, but `"foo"bar"` should not be.

Before this PR, the following TextMate rule is generated (simplified):

```
{
  "begin": "\""
  "end": "\""
  "name": "string"
  "patterns": [
    { "match": "[a-z]" },
    { "match": "[\\]" },
    { "match": "[\"]" }
  ]
}
```

That is, roughly, for each symbol `s` in `Char`, a nested pattern is included to gobble up content between `begin` and `end` that matches individually `s`. However, gobbling up individual symbols is too fine-grained. For instance, with this rule, `"foo\"bar"` will *not* be fully highlighted: only `"foo\"` will be highlighted (because the `\` is not considered in combination with the second `"`).

Instead of gobbling up individual symbols, the tokenizer should gobble up *segments* of symbols that belong together. That is what this PR adds.

After this PR, the following TextMate rule is generated (simplified):

```
{
  "begin": "\""
  "end": "\""
  "name": "string"
  "patterns": [
    { "match": "[a-z]" },
    { "match": "[\\][\"]" }
  ]
}
```

### CODE REVIEW

File: rascal-textmate-core/src/main/rascal/lang/rascal/grammar/Util.rsc [hunk 1]
```diff
@@ -31,15 +31,6 @@ bool tryParse(Grammar g, Symbol s, str input, bool allowAmbiguity = false) {
     return false;
 }
 
-@synopsis{
-    Gets the terminals that occur in production `p`, possibly recursively
-    (default: `true`)
-}
-
-set[Symbol] getTerminals(Grammar g, Production p, bool recur = true)
```
Comment (sungshik) [inline rascal-textmate-core/src/main/rascal/lang/rascal/grammar/Util.rsc:39]: This function was used before this PR to compute individual symbols to gobble up between `begin`/`end`, but it's not longer needed

File: rascal-textmate-core/src/main/rascal/lang/rascal/grammar/analyze/Symbols.rsc [hunk 2]
```diff
@@ -112,16 +113,6 @@ private map[Symbol, Maybe[set[Symbol]]] followBySymbol(Grammar g, bool(Symbol) p
     return ret;
 }
 
-private set[Symbol] unmaybe(just(set[Symbol] \set))
```
Comment (sungshik) [inline rascal-textmate-core/src/main/rascal/lang/rascal/grammar/analyze/Symbols.rsc:115]: Move the following functions to a new utility module, `util::MaybeUtil`

File: rascal-textmate-core/src/main/rascal/util/MaybeUtil.rsc [hunk 3]
```diff
@@ -0,0 +1,28 @@
+@synopsis{
```
Comment (sungshik) [inline rascal-textmate-core/src/main/rascal/util/MaybeUtil.rsc:1]: The functions in this module were previously part of module `lang::rascal::grammar::analyze::Symbols`, but they're now needed also in other places

**Review by sungshik — COMMENTED** [review]
A few additional comments

File: rascal-textmate-core/src/main/rascal/lang/rascal/grammar/analyze/Symbols.rsc (hunk 2 above)
Comment (DavyLandman) [reply rascal-textmate-core/src/main/rascal/lang/rascal/grammar/analyze/Symbols.rsc:115]: looks to me like it should always have been a ``set`? as the empy set is a fine representation of "nothing"?

File: rascal-textmate-core/src/main/rascal/util/MaybeUtil.rsc [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+@synopsis{
+    Returns just the union of the sets of two `Maybe` values when present.
+    Returns nothing if absent.
+}
+
+Maybe[set[&T]] union(just(set[&T] set1), just(set[&T] set2))
+    = just(set1 + set2);
+
+default Maybe[set[&T]] union(Maybe[set[&T]] _, Maybe[set[&T]] _)
```
Comment (DavyLandman) [inline rascal-textmate-core/src/main/rascal/util/MaybeUtil.rsc:27]: what if one of the two was a `just`?

File: vscode-extension/syntaxes/pico.tmLanguage.json [hunk 5]
```diff
@@ -14,7 +14,15 @@
...
-          "match": "((?:\\\")[\\x{01}-\\!\\#-\\x{10FFFF}]*?(?:\\\"))",
+          "match": "((?:\\\\)[\\\"\\\\])",
+          "captures": {
+            "1": {
+              "name": "string.quoted.double"
+            }
+          }
+        },
+        {
+          "match": "([\\x{01}-\\!\\#-\\[\\]-\\x{10FFFF}])",
```
Comment (DavyLandman) [inline vscode-extension/syntaxes/pico.tmLanguage.json:25]: in the rascal grammar this is written as `![\\"]` right? can we not generate a negative regex class? `[^\\"]`?

**Review by DavyLandman — APPROVED** [review]
This looks a lot better indeed :+1: also the textmate regexes look more like you would expect them too.

**Review by sungshik — COMMENTED** [review]
Thank you for the quick review, @DavyLandman, 🙂

**Pull request merged**